src/solver/pooling/backward2d.cpp

```diff
@@ -0,0 +1,206 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/pooling/solvers.hpp>
+
+#include <miopen/pooling/invoke_params.hpp>
+#include <miopen/pooling/problem_description.hpp>
+#include <miopen/datatype.hpp>
+#include <miopen/pooling.hpp>
+#include <miopen/kernel_build_params.hpp>
+
+namespace miopen {
+
+namespace solver {
+
+namespace pooling {
+
+bool PoolingBackward2d::IsApplicable(const ExecutionContext&,
+                                     const miopen::pooling::ProblemDescription& problem) const
+{
+    return problem.GetDirection() == miopen::pooling::Direction::Backward &&
+           (problem.GetPooling().GetMode() == miopenPoolingMax ||
+            problem.GetPooling().GetMode() == miopenPoolingAverage ||
+            problem.GetPooling().GetMode() == miopenPoolingAverageInclusive) &&
+           problem.GetXDesc().GetSize() == 4;
+}
+
+ConvSolution
+PoolingBackward2d::GetSolution(const ExecutionContext&,
+                               const miopen::pooling::ProblemDescription& problem) const
+{
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    {
+        auto kernel = KernelInfo{};
+
+        kernel.kernel_file = "MIOpenPoolingBwd.cl";
+
+        if(problem.GetPooling().GetMode() == miopenPoolingMax)
+        {
+            kernel.kernel_name = "mloPoolingMaxBwd";
+        }
+        else if(problem.GetPooling().GetMode() == miopenPoolingAverage ||
+                problem.GetPooling().GetMode() == miopenPoolingAverageInclusive)
+        {
+            kernel.kernel_name = "mloPoolingAveBwd";
+        }
+
+        std::size_t batch_sz, n_inputs, in_height, in_width;
+        std::tie(batch_sz, n_inputs, in_height, in_width) =
+            miopen::tien<4>(problem.GetXDesc().GetLengths(), 1);
+
+        const int pooling_method = (problem.GetPooling().GetMode() == miopenPoolingMax)
+                                       ? MLO_POOLING_OP_MAX
+                                       : ((problem.GetPooling().GetMode() == miopenPoolingAverage)
+                                              ? MLO_POOLING_OP_AVE
+                                              : MLO_POOLING_OP_AVE_INCLUSIVE);
+
+        std::size_t grp_tile0 = 8;
+        std::size_t grp_tile1 = 8;
+
+        std::size_t out_pix_tile0 = 1;
+        std::size_t out_pix_tile1 = 1;
+
+        if(problem.GetPooling().GetMode() == miopenPoolingMax)
+        {
+            // clang-format off
+            grp_tile0 =
+                   in_width <= 8   ? 8
+                : (in_width <= 16  ? 4
+                : (in_width <= 24  ? 8
+                : (in_width <= 32  ? 32
+                : (in_width <= 64  ? 8
+                : (in_width <= 96  ? 16
+                : (in_width <= 128 ? 16
+                                   : 32))))));
```

Comment:
It's better to add a comment to clarify what's going on here

Reply:
@averinevg This is ascii art, I guess ;)

Seriously, this code existed before this PR. This change simply moves it to another place. You can ask the original author of this code to add comments (you need to open a separate ticket for that).

Reply:
I have also formatted such pieces to be less of an ascii art and more of an understandable code. But otherwise, it has only been moved.

Reply:
@DrizztDoUrden Please do not destroy art for the sake of utility! :rofl:

Reply:
🤣 any improvements to readability are definitely welcomed!